internal/manifest/oci_index_test.go

```diff
@@ -45,3 +49,113 @@ func TestOCI1IndexFromManifest(t *testing.T) {
+
+func TestOCI1IndexChooseInstanceByCompressionAndListInternalFromBlob(t *testing.T) {
+	type expectedMatch struct {
+		arch, variant  string
+		instanceDigest digest.Digest
+	}
+	for _, manifestList := range []struct {
+		listFile               string
+		matchedInstances       []expectedMatch
+		prefferedGzipInstances []expectedMatch
+		unmatchedInstances     []string
+	}{
+		{
+			listFile: "oci1.index.zstd-selection.json",
+			matchedInstances: []expectedMatch{
+				// out of gzip and zstd in amd64 select the first zstd image
+				{"amd64", "", "sha256:ac9bbf71dacfad2bfbe48f7d46ca279b1b3b351a6c814db760030015167405b8"},
+				// out of multiple gzip in arm64 select the first one to ensure original logic is prevented
+				{"arm64", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74a9522511"},
+				// select a signle gzip s390x image
+				{"s390x", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74z9522511"},
+			},
+			unmatchedInstances: []string{
+				"unmatched",
+			},
+			// When preffredGzip: true
+			prefferedGzipInstances: []expectedMatch{
+				// out of gzip and zstd in amd64 select the first zstd image
+				{"amd64", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74a9522510"},
+				// out of multiple gzip in arm64 select the first one to ensure original logic is prevented
+				{"arm64", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74a9522511"},
+				// select a signle gzip s390x image
+				{"s390x", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74z9522511"},
+			},
+		},
+		{
+			listFile: "oci1.index.zstd-selection2.json",
+			// out of list where first instance is gzip , select the first occurance of zstd out of many
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:ac9bbf71dacfad2bfbe48f7d46ca279b1b3b351a6c814db760030015167405b8"},
+			},
+			prefferedGzipInstances: []expectedMatch{
+				{"amd64", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74a9522510"},
+			},
+			unmatchedInstances: []string{
+				"unmatched",
+			},
+		},
+		{ // Focus on ARM variant field testing
+			listFile: "schema2list-variants.json",
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:59eec8837a4d942cc19a52b8c09ea75121acc38114a2c68b98983ce9356b8610"},
+				{"arm", "v7", "sha256:f365626a556e58189fc21d099fc64603db0f440bff07f77c740989515c544a39"},
+				{"arm", "v6", "sha256:f365626a556e58189fc21d099fc64603db0f440bff07f77c740989515c544a39"},
+				{"arm", "v5", "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"},
+				{"arm", "", "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"},
+				{"arm", "unrecognized-present", "sha256:bcf9771c0b505e68c65440474179592ffdfa98790eb54ffbf129969c5e429990"},
+				{"arm", "unrecognized-not-present", "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"},
+			},
+			unmatchedInstances: []string{
+				"unmatched",
+			},
```

Comment:
This is not relevant to the function under test at all.

Reply:
… but what _would_ be relevant is doing a similar variant preference test against OCI, and making sure that it interacts _correctly_ with the compression preference — i.e. that the variant preference is more important.

Reply:
Added a test for this.